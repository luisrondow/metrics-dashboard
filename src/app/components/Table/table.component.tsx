import { StyledTable, StyledThead, StyledTh, StyledTd, TableWrapper } from './table.styles';

type TableProps<T> = {
  columns: Array<keyof T>;
  data: T[];
};

const Table = <T,>({ columns, data }: TableProps<T>) => {
  return (
    <TableWrapper>
      <StyledTable>
        <StyledThead>
          <tr>
            {columns.map((column) => (
              <StyledTh key={String(column)}>{String(column)}</StyledTh>
            ))}
          </tr>
        </StyledThead>
        <tbody>
          {data.map((row, rowIndex) => (
            <tr key={rowIndex}>
              {columns.map((column) => (
                <StyledTd key={`${rowIndex}-${String(column)}`}>{String(row[column])}</StyledTd>
              ))}
            </tr>
          ))}
        </tbody>
      </StyledTable>
    </TableWrapper>
  );
};

export default Table;
