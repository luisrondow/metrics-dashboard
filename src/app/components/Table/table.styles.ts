import styled from 'styled-components';

export const TableWrapper = styled.div`
  overflow-x: auto;
  margin: 20px 0;
`;

export const StyledTable = styled.table`
  width: 100%;
  border-collapse: collapse;
`;

export const StyledThead = styled.thead`
  background-color: #f5f5f5;
`;

export const StyledTh = styled.th`
  padding: 10px;
  border: 1px solid #ddd;
  text-align: left;
`;

export const StyledTd = styled.td`
  padding: 10px;
  border: 1px solid #ddd;
`;
