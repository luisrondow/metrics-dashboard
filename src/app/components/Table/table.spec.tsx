import { render, screen } from '@testing-library/react';
import Table from './table.component';

jest.mock('./table.styles', () => {
  const TableWrapper = ({ children }: { children: React.ReactNode }) => <div>{children}</div>;

  const StyledTable = ({ children }: { children: React.ReactNode }) => <table>{children}</table>;

  const StyledThead = ({ children }: { children: React.ReactNode }) => <thead>{children}</thead>;

  const StyledTh = ({ children }: { children: React.ReactNode }) => <th>{children}</th>;

  const StyledTd = ({ children }: { children: React.ReactNode }) => <td>{children}</td>;

  return { TableWrapper, StyledTable, StyledThead, StyledTh, StyledTd };
});

describe('Table Component', () => {
  type TestData = {
    Name: string;
    Age: number;
    Address: string;
  };

  const columns: Array<keyof TestData> = ['Name', 'Age', 'Address'];
  const data: TestData[] = [
    { Name: 'Jose da Silva', Age: 28, Address: 'Rua de Camoes' },
    { Name: 'Joao Pedro', Age: 34, Address: 'Rua de Faria Guimaraes' },
    { Name: 'Ines Tavares', Age: 22, Address: 'Rua de Costa Cabral' },
  ];

  it('matches snapshot', () => {
    const { asFragment } = render(<Table columns={columns} data={data} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders table with columns', () => {
    render(<Table columns={columns} data={data} />);
    columns.forEach((column) => {
      expect(screen.getByText(String(column))).toBeInTheDocument();
    });
  });

  it('renders table with data', () => {
    render(<Table columns={columns} data={data} />);
    data.forEach((row) => {
      columns.forEach((column) => {
        expect(screen.getByText(String(row[column]))).toBeInTheDocument();
      });
    });
  });

  it('renders correct number of rows and columns', () => {
    render(<Table columns={columns} data={data} />);
    const rows = screen.getAllByRole('row');
    expect(rows).toHaveLength(data.length + 1); // including header row

    const cells = screen.getAllByRole('cell');
    expect(cells).toHaveLength(data.length * columns.length);
  });
});
