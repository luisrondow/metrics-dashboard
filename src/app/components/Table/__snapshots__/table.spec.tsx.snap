// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <table>
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Age
          </th>
          <th>
            Address
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Jose da Silva
          </td>
          <td>
            28
          </td>
          <td>
            Rua de Camoes
          </td>
        </tr>
        <tr>
          <td>
            Joao Pedro
          </td>
          <td>
            34
          </td>
          <td>
            Rua de Faria Guimaraes
          </td>
        </tr>
        <tr>
          <td>
            Ines Tavares
          </td>
          <td>
            22
          </td>
          <td>
            Rua de Costa Cabral
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
